packages/boot/test/smartWallet/wallet-fun.test.ts

```diff
@@ -0,0 +1,413 @@
+      const update = wd.getLatestUpdateRecord();
+      assert.equal(update.updated, 'offerStatus');
+      t.log('Seller payouts', update.status.payouts);
+      t.like(update.status.payouts, {
+        Price: { value: Price.value },
+      });
+    },
+  });
+};
+
+test('use Invitation offer result in atomicSwap', async t => {
+  const { walletFactoryDriver, agoricNamesRemotes } = t.context;
+  const addr = { A: 'agoric1partyA', B: 'agoric1partyB' };
+  const wd = {
+    A: await walletFactoryDriver.provideSmartWallet(addr.A),
+    B: await walletFactoryDriver.provideSmartWallet(addr.B),
+  };
+
+  {
+    // in kernel world...
+    const { EV } = t.context.runUtils;
+    const agoricNames = await EV.vat('bootstrap').consumeItem('agoricNames');
+    const issuerHub = await EV(agoricNames).lookup('issuer');
+    const issuers = {
+      Asset: await EV(issuerHub).lookup('ATOM'),
+      Price: await EV(issuerHub).lookup('BLD'),
+    };
+    // XXX seller should start contract
+    const swap = await startContract(
+      t,
+      startSwap as ContractStartFn,
+      '@agoric/zoe/src/contracts/atomicSwap.js',
+      { name: 'atomicSwap', issuers },
+    );
+    const namesByAddress =
+      await EV.vat('bootstrap').consumeItem('namesByAddress');
+    const depositA = await EV(namesByAddress).lookup(addr.A, 'depositFacet');
+    await EV(depositA).receive(swap.creatorInvitation);
+  }
+
+  const { brand } = agoricNamesRemotes;
+  const Asset = AmountMath.make(brand.ATOM as unknown as Brand, 2n);
+  const Price = AmountMath.make(brand.BLD as unknown as Brand, 30n);
+  // client-side presence
+  const { atomicSwap } = agoricNamesRemotes.instance;
+  t.truthy(atomicSwap);
+  const party = {
+    A: makeTrader(addr.A, wd.A, atomicSwap, { Asset, Price }),
+    B: makeTrader(addr.B, wd.B, atomicSwap, { Asset, Price }),
+  };
+
+  await party.A.startSale(t, addr.B);
+  await party.B.buy(t);
+  await party.A.completeSale(t);
+});
+
+const makeUpgradeAdmin = (
+  wd: WalletDriver,
+  carrier: Instance<typeof startCarrierContract>,
+) => {
+  return harden({
+    async redeem(t: ExecutionContext) {
+      await wd.executeOffer({
+        id: 'redeemInvitation',
+        invitationSpec: {
+          source: 'purse',
+          description: 'admin',
+          instance: carrier,
+        },
+        proposal: {},
+        after: { saveAs: 'kit' },
+      });
+      return wd.getLatestUpdateRecord();
+    },
+    async restart(t: ExecutionContext) {
+      const restartFn =
+        (
+          E = null as unknown as EProxy,
+          my = {} as { kit: StartedInstanceKit<typeof startPriceContract> },
+        ) =>
+        () =>
+          E(
+            after(my.kit.publicFacet, {
+              rx: my.kit.adminFacet as any, // XXX after() type
+              method: 'restartContract',
+              args: [my.kit.privateArgs],
+            }),
+          ).getIncarnation(); // use the public (TODO: creator) facet
```

Comment:
I learned in discussion with @kriskowal that this is needlessly contorted; it could/should be:

`after(my.kit.publicFacet, E(my.kit.adminFacet).restartContract(my.kit.privateArgs)).getIncarnation()`

with...

```ts
const after = (obj, prereq) = E.when(prereq, () => obj)
```